### PR TITLE
QA-SISRP-18744 - Advisor student overview page needs a better UX while data elements are loading

### PR DESCRIPTION
**Pull request by dfkaye**
https://jira.berkeley.edu/browse/SISRP-18744

### CODE REVIEW

File: src/assets/javascripts/angular/controllers/pages/userOverviewController.js [hunk 1]
```diff
@@ -79,9 +79,9 @@ angular.module('calcentral.controllers').controller('UserOverviewController', fu
   $scope.$on('calcentral.api.user.isAuthenticated', function(event, isAuthenticated) {
     if (isAuthenticated) {
       // Refresh user properties because the actAsOptions.canSeeCSLinks property is sensitive to the /user/overview route.
-      apiService.user.fetch();
-      loadProfile();
-      loadAcademics();
+      apiService.user.fetch()
+      .then(loadProfile)
+      .then(loadAcademics);
```
Comment (johncrossman) [inline src/assets/javascripts/angular/controllers/pages/userOverviewController.js]: @dfkaye @davidlimcheng You each have an open PR that touches this block in `userOverviewController.js`. If they're both merged without conflict we'll need a follow up PR to have:
```
.then(loadHolds)
etc.
```

Comment (johncrossman) [inline src/assets/javascripts/angular/controllers/pages/userOverviewController.js]: @dfkaye The other PR has been closed so my statement above is no longer a concern.

File: src/assets/templates/widgets/enrollment_card.html [hunk 2]
```diff
@@ -1,40 +1,46 @@
-<div data-ng-controller="EnrollmentCardController" data-cc-spinner-directive="enrollment.isLoading" class="cc-enrollment-card">
-  <div class="cc-widget cc-widget-enrollment-card" data-ng-repeat="enrollmentTerm in enrollment.terms" data-ng-if="enrollmentTerm.enrollmentPeriod">
+<div class="cc-enrollment-card" data-ng-controller="EnrollmentCardController">
+  <div class="cc-widget" data-ng-if="enrollment.isLoading">
```
Comment (johncrossman) [inline src/assets/templates/widgets/enrollment_card.html]: @dfkaye Users with no enrollments will see card disappear after a few seconds of spinner. This seems wrong – we need some UX guidance. Do we want to keep the card in place and say "No enrollments"?

Comment (johncrossman) [inline src/assets/templates/widgets/enrollment_card.html]: @dfkaye @raydavis This change has already been merged to master. See #5364. We might need to patch master.

Comment (raydavis) [inline src/assets/templates/widgets/enrollment_card.html]: Argh! Thank you for catching that, @johncrossman , and I'm sorry I didn't. You're right that to preserve the current UX design we must keep the spinner directive on the outer div.

Comment (raydavis) [inline src/assets/templates/widgets/enrollment_card.html]: Although of course what https://jira.berkeley.edu/browse/SISRP-18744 complained about was the "current UX".

@pfarestveit , as you wrote in the ticket, "we have data dependencies such that we can't be sure exactly which cards will show up or not until we have certain other data in place." Unless we have a solution already in place for the student My Academics view, this is a difficult UX question which should involve the designers.

Comment (dfkaye) [inline src/assets/templates/widgets/enrollment_card.html]: @raydavis @johncrossman Give me a few minutes to iron this one out

Comment (dfkaye) [inline src/assets/templates/widgets/enrollment_card.html]: @raydavis @johncrossman @pfarestveit 
We can add one more check for `enrollment.terms.length > 0`, like so:

    <div class="cc-widget" data-ng-if="enrollment.isLoading && enrollment.terms.length > 0">

Effects:
+ The default case (no terms yet) means the enrollment card and spinner will not be displayed (no more peek-a-boo).
+ Only after the terms request completes and there are terms in the array, will the card then appear with the spinner until each term is processed (which means the card appears in loading state *after* student loads.

Shall we try that?

Comment (raydavis) [inline src/assets/templates/widgets/enrollment_card.html]: @dfkaye , as I commented at the JIRA ticket, we _probably_ want to match the existing student My Academics approach on this...

Comment (dfkaye) [inline src/assets/templates/widgets/enrollment_card.html]: @raydavis how about a "No Enrollments" message then? That way the spinner is up and replaced either by the message or by each enrollment card

      <div class="cc-widget" data-ng-if="enrollment.isLoading">
        <div class="cc-widget-title">
          <h2>Class Enrollment</h2>
        </div>
        <div data-cc-spinner-directive="enrollment.isLoading"></div>
      </div>
      <div class="cc-widget" data-ng-if="!enrollment.isLoading">
        <div class="cc-widget-title">
          <h2>Class Enrollment</h2>
        </div>
        <div class="cc-widget-text">No enrollments.</div>
      </div>
      <div data-ng-if="!enrollment.isLoading">

         show the enrollment cards....

Comment (johncrossman) [inline src/assets/templates/widgets/enrollment_card.html]: @dfkaye The 'No Enrollments' message for *all users* might be a problem. We should hide the card if it's obviously irrelevant to the user (e.g., instructor).

Comment (raydavis) [inline src/assets/templates/widgets/enrollment_card.html]: @dfkaye , @johncrossman , the existing student My Academics UX does not show a "Class Enrollment Fall 2016" card for students without a Fall 2016 enrollment term in CS, and so I wouldn't think we'd need to show one to advisors either.

Comment (dfkaye) [inline src/assets/templates/widgets/enrollment_card.html]: @raydavis @johncrossman We cannot match the existing My Academics experience.  Either we show "No enrollments" after the spinner, or we show the spinner once the `enrollment.terms` array is populated.

**Comment by johncrossman** [comment]
@dfkaye @raydavis Unfortunately, we have no easy answer. I'd like to put off this discussion and drop the proposed changes to `enrollment_card.html`. Of course, that means reverting PR #5364 on master.

**However...**  I would like the change to `userOverviewController.js` on both master and QA. 

I apologize for the back and forth, @dfkaye  Let's prioritize this in the next sprint and devote this week to our release-blockers.

**Comment by dfkaye** [comment]
@johncrossman I'm closing this PR (it's a QA PR anyway) ~ we'll re-visit soon

**Pull request closed without merging**